### ISSUE TITLE
Reading certificate fails on mcr.microsoft.com/dotnet/core/runtime:2.1-nanoserver-1803 and later

### ISSUE THREAD

**Issue by dozer75** [dotnet/dotnet-docker#1048]
For some reason it seems that reading certificates from files using nanoserver images newer or equal to at least 1803 fails with exception as described in Actual behavior. This applies both to runtime and aspnet images.

Linux and Nanoserver-sac2016 works as expected and running it outside of Docker.

## Steps to reproduce the issue
Either run `docker run -it --rm dozer75/demo:certtest-1809` and `docker run -it --rm dozer75/demo:certtest-latest` to see the exception, or build your own examples using `https://github.com/dozer75/certtest`

Note: Depending on your environment you may need to run containers with hyperv isolation.

## Expected behavior
The loading of certificate from a file should work as it does with `docker run -it --rm dozer75/demo:certtest-2016` and `docker run -it --rm dozer75/demo:certtest-linux`.

## Actual behavior
Loading certificate fails with the stacktrace:
`Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: Object was not found
   at Internal.Cryptography.Pal.CertificatePal.FilterPFXStore(Byte[] rawData, SafePasswordHandle password, PfxCertStoreF
lags pfxCertStoreFlags)
   at Internal.Cryptography.Pal.CertificatePal.FromBlobOrFile(Byte[] rawData, String fileName, SafePasswordHandle passwo
rd, X509KeyStorageFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate..ctor(String fileName, String password, X509KeyStora
geFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate2..ctor(String fileName, String password)
   at certtest.Program.Main(String[] args) in C:\Development\certtest\certtest\Program.cs:line 14`

## Additional information (e.g. issue happens only occasionally)
Tried to search on the error `Object was not found` but did only get solutions around IIS and user profile. Wondering if it is some invalid user profile on the base packages?

This issue causes file certificates that is sent to Kestrel using the `ASPNETCORE_Kestrel__Certificates__Default__Path` environment variable to fail with the same exception and thus impossible to use file certificates HTTPS and Kestrel on nanoserver.

## Output of `docker version`

```
Client: Docker Engine - Community
 Version:           18.09.2
 API version:       1.39
 Go version:        go1.10.8
 Git commit:        6247962
 Built:             Sun Feb 10 04:12:31 2019
 OS/Arch:           windows/amd64
 Experimental:      false

Server: Docker Engine - Community
 Engine:
  Version:          18.09.2
  API version:      1.39 (minimum version 1.24)
  Go version:       go1.10.6
  Git commit:       6247962
  Built:            Sun Feb 10 04:28:48 2019
  OS/Arch:          windows/amd64
  Experimental:     false
```

Also fails on other environments

## Output of `docker info`

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 64
Server Version: 18.09.2
Storage Driver: windowsfilter
 Windows:
Logging Driver: json-file
Plugins:
 Volume: local
 Network: ics l2bridge l2tunnel nat null overlay transparent
 Log: awslogs etwlogs fluentd gelf json-file local logentries splunk syslog
Swarm: inactive
Default Isolation: hyperv
Kernel Version: 10.0 17763 (17763.1.amd64fre.rs5_release.180914-1434)
Operating System: Windows 10 Enterprise Version 1809 (OS Build 17763.379)
OSType: windows
Architecture: x86_64
CPUs: 8
Total Memory: 31.82GiB
Name: XXXXXXXXXX
ID: EK4Y:324D:AGMO:A6TP:3VOV:P5SU:MCVB:FASP:TZSM:Y5W2:NBZX:YDSA
Docker Root Dir: C:\ProgramData\Docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: -1
 Goroutines: 54
 System Time: 2019-04-10T22:51:15.4095305+02:00
 EventsListeners: 1
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false
Product License: Community Engine
```

Also fails on other environments

**Comment by dozer75** [comment]
Found #915 today, it seems to be a similar issue but with a different error, it works with older images of nanoserver, but not the later ones. I have seen that exception once too, but mostly the exception as this issue states. This issue contains a smaller reproducible example than using aspnetsamples.

**Comment by MichaelSimons** [comment]
@dozer75 - I believe the issue you are encountering stems from the fact that in Nano Server versions after 2016 the default user is the `ContainerUser` versus the `ContainerAdministrator`.  If you run your `1803` image with the `ContainerAdministrator` (e.g. `docker run --user ContainerAdministrator...`) you will see it works as expected.

You can get this to work with the `ContainerUser` by changing the `X509KeyStorageFlags` (e.g. X509KeyStorageFlags.EphemeralKeySet).  I've asked @bartonjs for some guidance on if/how to configure the `ContainerUser` to work with the default storage option.

**Comment by dozer75** [comment]
@MichaelSimons - Both solutions works for my example. However as mentioned in my issue and in #915 this causes issues with ASP.NET Core and Kestrel (they are probably not using `X509KeyStorageFlags.EphemeralKeySet`)  and thus it is a requirement to run it as `ContainerAdministrator`, so at least the ASP.NET Core images should be adjusted to have `ContainerAdministrator` as default since it is a requirement to get HTTPS to work.

Anyway, I think that it shouldn't be necessary to run the package as `ContainerAdministrator` to use certificates, I suspect that the reason why `ContainerUser` and `ContainerAdministrator` exists in first place to limit access due to security related issues and it feels like a bad idea to run Kestrel as `ContainerAdministrator`. I also have a feeling that this is the same issue that appears in IIS when using certificates with an app pool that doesn't have "Load User Profile" set (that `ContainerUser` is lacking a user profile?). It works in the Linux environment without any adjustments, so it should be the same with the nano server packages.

**Comment by bartonjs** [comment]
The best answer I have is that ContainerUser doesn't seem to get a user profile, therefore it cannot have persisted keys (or "temporary" persisted keys).

I don't know enough to say if it can be made to load one, or if a custom user can be made to get one, or what.

**Comment by robertmuehsig** [comment]
Just want to point out, that this is still an issue with:

    mcr.microsoft.com/dotnet/core/aspnet:3.1-nanoserver-1809 

Not sure if I miss something, but running it with ContainerAdministrator (which sounds bad), seems to fix it for now.

**Comment by nickwesselman** [comment]
Seeing the same issue with

```
mcr.microsoft.com/dotnet/aspnet:5.0
```

**Comment by bartonjs** [comment]
This is really a Windows issue, or maybe a container configuration issue.  To load a PFX on Windows you need to do one of

* Use X509KeyStorageFlags.EphemeralKeySet
* Be a user with a loaded user profile (this is the part that isn't true on default Windows containers)
* Use X509KeyStorageFlags.MachineKeySet (and have appropriate permissions to do so)

(Ignoring that some PFXes will try to load into the machine keyset by default)

**Comment by nickwesselman** [comment]
Running as `ContainerAdministrator` on the latest `mcr.microsoft.com/dotnet/aspnet:5.0` doesn't seem to work any more since I pulled the latest. 🙁

**Comment by scottdcarson** [comment]
TL;DR: you can't import a pfx certificate into the user store if it was originally exported from a machine store.  Import it into your user store, export it as a pfx, then make it available to your nanoserver container (typically as a secret).

It's not a bug, it's just something to understand (and with a painful learning curve).

Using X509KeyStorageFlags.EphemeralKeySet is a non-starter, because then you won't be able to access the private key later.  The user profile thing is a red herring, reminiscent of the days when IIS_APPPool users didn't have the "create user profile" setting - but it doesn't apply here.  And using X509KeyStorageFlags.MachineKeySet is also a nonstarter, because you'd only be able to do that as ContainerAdministrator, and you'd only be able to access the private key as ContainerAdministrator, and you should run your .net core web server as ContainerUser.

Here's what works.  Start by completing a certificate request, making sure that you request that the private key be exportable.  Either initiate the request from your user store, or export the pfx and re-import it into your user store.  Then export from your user (not machine) store, including the private key.  Make the resulting pfx file available to your nanoserver container (perhaps as a k8s or docker swarm secret).  Here's a helper class that I use:

```
    public static class X509Certificate2Factory
    {
        /// <summary>
        /// Generate an X509Certificate2 from .pfx certificate Data and password.  BE SURE THE CERTIFICATE WAS EXPORTED FROM THE USER STORE!
        /// </summary>
        /// <param name="certificateData"></param>
        /// <param name="password"></param>
        /// <returns></returns>
        public static X509Certificate2 FromSecretData(byte[] certificateData, string password)
        {
            // remove any trailing newline that may have been inserted by a container orchestrator
            var certificatePassword = password.TrimEnd(Environment.NewLine.ToCharArray()).Trim();
            try
            {
                var certificate = new X509Certificate2(certificateData, certificatePassword, X509KeyStorageFlags.Exportable);
                return certificate;
            }
            catch (Exception ex)
            {
                if (Regex.IsMatch(ex.Message, ".*object.*not.*found.*", RegexOptions.IgnoreCase))
                {
                    // this probably means that the certificate was exported from the machine store, and is trying
                    // unsuccessfully to import itself back there when the user doesn't have permission
                    throw new ArgumentException($"Failed to import the certificate - be sure it was exported from the USER, not MACHINE store ({ex.GetType()})", "certificateData");
                }
                else
                {
                    throw; // some other error
                }
            }
        }
    }
```

Note that you'll get the obscure "object was not found" error if you try to import a certificate that was exported from the machine store.

**Comment by bartonjs** [comment]
@scottdcarson You should be able to force it into the user store with X509KeyStorageFlags.UserKeySet.  (That effectively says "ignore that the PFX says the key should go to the machine store")

**Comment by scottdcarson** [comment]
@bartonjs, thanks for the tip, I tested and it worked.  I believe we have successfully collaborated.

**Comment by nicholasyin** [comment]
When can we expect to skip this step? It's not mentioned in any official documentation. Only came here from search result using the exception text.

**Comment by mloskot** [comment]
Having this in `appsettings.json`

```json
  "Kestrel": {
    "Endpoints": {
      "HttpsInlineCertFile": {
        "Url": "https://localhost:8081",
        "Certificate": {
          "Path": "./localhost.pfx",
          "Password": "secret123"
        }
      }
    },
```

It is impossible to start server due to

```
Internal.Cryptography.CryptoThrowHelper+WindowsCryptographicException: Object was not found.
   at Internal.Cryptography.Pal.CertificatePal.FilterPFXStore(ReadOnlySpan`1 rawData, SafePasswordHandle password, PfxCertStoreFlags pfxCertStoreFlags)
   at Internal.Cryptography.Pal.CertificatePal.FromBlobOrFile(ReadOnlySpan`1 rawData, String fileName, SafePasswordHandle password, X509KeyStorageFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate..ctor(String fileName, String password, X509KeyStorageFlags keyStorageFlags)
   at System.Security.Cryptography.X509Certificates.X509Certificate2..ctor(String fileName, String password)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Certificates.CertificateConfigLoader.LoadCertificate(CertificateConfig certInfo, String endpointName)
   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.Reload()
   at Microsoft.AspNetCore.Server.Kestrel.KestrelConfigurationLoader.Load()
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.BindAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.KestrelServerImpl.StartAsync[TContext](IHttpApplication`1 application, CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Hosting.GenericWebHostService.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.Internal.Host.StartAsync(CancellationToken cancellationToken)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.RunAsync(IHost host, CancellationToken token)
   at Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run(IHost host)
   at <...snip...>.Program.Main(String[] args) in C:\<...snip...>\Program.cs:line 38
```

unless NanoServer-based container like `mcr.microsoft.com/dotnet/aspnet:6.0` is `docker run`-ed with `--user ContainerAdministrator`

The same requirement applies to use the now deprecated `certoc.exe` utility to import PFX on the NanoServer-based container which, without being run with  `--user ContainerAdministrator` fails with `0x80090011 (-2146893807 NTE_NOT_FOUND)` error.
The `certoc.exe -ImportPFX -p secret123 Root .\localhost.pfx` succeeds if container runs with `--user ContainerAdministrator`.

**Comment by dozer75** [comment]
I am the creator of this issue, and I'll close it as well. 

Microsoft seems reluctant to fix an issue that is obviously wrong (but they have mentioned in a [sample](https://github.com/dotnet/dotnet-docker/blob/main/samples/host-aspnetcore-https.md) (it states that it's only related to selv-signed certificates, but this is not true))  so there is no point to keep it open, so this is probably the best solution we'll get on this issue.

There are some workarounds mentioned in this issue that may  help others. I have completely stopped using Windows containers due to their nature and seemingly uninterest from Microsoft to fix obvious problems for Windows Containers and the fact that Linux based containers and hosting environments works better overall.

The workarounds are:
- Use ContainerAdministrator (confirmed to work)
- Use Scotts solution (probably works, but not tried it)

If others want to take this further rather create a new issue and reference this as well.

**Comment by jonproch117** [comment]
New to the procedure, so if this should go elsewhere I'm happy to move it

After encountering this issue myself, I believe I've found a way to get this to consistently work while running as ContainerUser, and without needing to do anything custom in the code to load the certificate. It appears the issue is with the way the PFX file itself is generated.

If I create the PFX file by exporting a certificate from the machine-wide Windows Cert Store, I get the error described above. However, if I create a certificate that is identical in every way, except that it was exported from the current user's Cert Store, Kestrel will launch successfully.

Upon digging around for any differences between the PFX files, I noticed that the one that came from the machine store actually has the fact that it is a _machine_ keyset encoded into the PFX file itself. This can be seen by running "certutil -asn <PFXFile>", and is somewhere in the middle of the hex dump, like so (the presence of the szOID_LOCAL_MACHINE_KEYSET is the important part):

```
04f8:    |           |              |        2f 51 38 5e fe 77 84 48  9c 4f 83 59 4b 36 0b c4  ; /Q8^.w.H.O.YK6..
0508:    |           |              |        19 e3 ae ca 31 73 60 c1  e3 8d 6c 43 c3 69 45 76  ; ....1s`...lC.iEv
0518:    |           |              |        54 ed e3 a8 51 a3 f2 b4  9f 8d 80 f1 76 35 1c 40  ; T...Q.......v5.@
0528:    |           |              |        43 1c c0 65 18 90 70 4f  61 1d 25 16 9f cb 3a e4  ; C..e..pOa.%...:.
0538:    |           |              |        34 19 b0 5c 74 c8 eb d5  18 c0 ba 2f db 4c a3 58  ; 4..\t....../.L.X
0548:    |           |              |        2d 60 89 40 fd 6e 36 de                           ; -`.@.n6.
0550:    |           |              31 81 e2    ; SET (e2 Bytes)
0553:    |           |                 30 0d    ; SEQUENCE (d Bytes)
0555:    |           |                 |  06 09 ; OBJECT_ID (9 Bytes)
0557:    |           |                 |  |  2b 06 01 04 01 82 37 11  02
         |           |                 |  |     ; 1.3.6.1.4.1.311.17.2 szOID_LOCAL_MACHINE_KEYSET
0560:    |           |                 |  31 00 ; SET (0 Bytes)
0562:    |           |                 30 13    ; SEQUENCE (13 Bytes)
0564:    |           |                 |  06 09 ; OBJECT_ID (9 Bytes)
0566:    |           |                 |  |  2a 86 48 86 f7 0d 01 09  15
         |           |                 |  |     ; 1.2.840.113549.1.9.21 szOID_PKCS_12_LOCAL_KEY_ID
056f:    |           |                 |  31 06 ; SET (6 Bytes)
```

PFX files that contain that entry end up requiring one of the workarounds mentioned above. But if you can generate your PFX file without it getting marked as a machine keyset, it should just work in your container by default. Exporting the certificate from the user store or generating it with openssl are 2 easy ways to get a PFX without the troublesome entry.

**Comment by scottdcarson** [comment]
@jonproch117  - This was mentioned above by myself and @bartonjs.  You can export from the user key store as you suggest, true, but you can also specify X509KeyStorageFlags.Exportable | X509KeyStorageFlags.UserKeySet in the constructor for X509Certificate2, which means "ignore the store from which the certificate came, and force it into the user store".  This works fine as ContainerUser, even if the certificate came from the machine store; we do it all the time.

**Comment by jonproch117** [comment]
@scottdcarson Thanks. I'd used that successfully to manually load the certificate previously. I guess I was thinking more in terms of the use case @mloskot posted where this is encountered when loading the Kestrel certificate from config, as I didn't see where/how the key storage flags can be set in that case. Maybe this is a different issue though.
